File: antia_numerical_methods/c/fun/bspfitw2.c
/*	To calculate linear least squares fit to B-spline basis functions in 2 dimension
	version for general weights, but much slower than BSPFIT2

	NX : (input) Number of data points along x-axis
	NY : (input) Number of data points along y-axis
	X,Y : (input) Array of length NX,NY containing the coordinates
		of points at which function values is available
	F : (input) Array of length IC*NY containing the function values
		F[j][i] should be function value at X[i], Y[j]
	EF : (input) Array of length IC*NY containing the error estimates in F[J][I]
	K : (input) Order of B-splines required, K=4 gives cubic B-splines
	A : (output) Array of length LA*3*NX*NY containing
		the matrix U of SVD of the design matrix
	LA : (input) Second dimension of array A as declared
		in the calling function  LA >= (MX+K-2)*(MY+K-2)
	V : (output) Array of length IV*(MX+K-2)*(MY+K-2) containing
		the matrix V of SVD of the design matrix
	IV : (input) Second dimension of V in the calling function
		IV >= (MX+K-2)*(MY+K-2)
	SIGMA : (output) Array of length (MX+K-2)*(MY+K-2)
		containing the singular values of the design matrix
	C : (output) Array of length IC*(MY+K-2) containing the fitted coefficients
	IC : (input) Second dimension of arrays C, F, EF, FY as declared
		in the calling function (IC >= NX)
	XF : (input) Array of size MX, containing the knots
		along x-axis used for defining B-spline basis functions.
	YF : (input) Array of size MY, containing the knots
		along y-axis used for defining B-spline basis functions.
	MX : (input) Number of knots for B-splines along x-axis,
		the number of basis functions would be MX+K-2
	MY : (input) Number of knots for B-splines along y-axis,
		the number of basis functions would be MY+K-2
	FY : (output) Array of length IC*NY containing the values of fitted
		function at each of the tabular points
	REPS : (input) Required accuracy for solution of equations using SVD
		singular values less than REPS times maximum will be set to zero
	RLM : (input) Parameter lambda for smoothing. If RLM<=0 no smoothing
		is applied
	IDE : (input) Order of derivative to be used for smoothing
		This is used only when RLM>0. IDE=1 for first derivative
		and IDE=2 for second derivative smoothing
	CHISQ : (output) The value of Chi square at minimum

	Error status is returned by the value of the function BSPFITW2.
		0 value implies successful execution
		608 implies that MX+K-2>NX, MY+K-2>NY or K<2
		609 implies that RLM>0 and IDE is not acceptable
		610 implies that EF[I][J]<=0 for some I,J
		No calculations are done in all these cases
		Other values may be set by SVD or BSPLIN

	Required functions : BSPLIN, BSPEV2, SVD, SVDEVL
*/

#include <math.h>
#include <stdlib.h>

int svdevl(int n, int m, double *u, double *v, double sigma[], int lu,
	int lv, double b[], double reps);
double bspev2(int nx, int ny, double *x, double *y, int k, int nderiv,
	double *wt, int iw, double x0, double y0, double *dfx, double *dfy,
	double *dfxx, double *dfyy, double *dfxy, int *ier);
int svd(int n, int m, double *a, double *v, double sigma[], int la, int lv);
int bsplin(double *x, int nx, int k, double xb, int nderiv, double b[],
		double db[], double ddb[], int *left);
double bspevl(int n, double *x, int k, int nderiv, double wt[], double x0,
		double *df, double *ddf, int *ier);

int bspfitw2(int nx, int ny, double x[], double y[], double *f, double *ef,
	int k, double *a, int la, double *v, int iv, double sigma[], double *c,
	int ic, double xf[], double yf[], int mx, int my, double *fy,
	double reps, double rlm, int ide, double *chisq)

{
	int i,j,n1,m1,m2,m,ndb,nderiv,nb,left,ier,k1,k2,ij,ij1;
	double xb,yb,r1,dfx,dfy,dfxx,dfxy,dfyy;
	double *wk, *wk1;

	if(nx<k || ny<k || k<2) return 608;

	if(rlm>0.0 && (ide<1 || ide>2)) return 609;

/*     N1 is the number of equations to be solved */
	n1=nx*ny;
	if(rlm>0.0) n1=3*n1;
 
/*     Set up the matrix equation and obtain SVD 
       M is the number of coefficients to be determined */
	m1=mx+k-2; m2=my+k-2; m=m1*m2;
	ndb=mx+4; if(my>mx) ndb=my+4;
	nderiv=0;
	if(rlm>0) {
		nderiv=ide;
		nb=ndb; if(ide==2) nb=2*ndb;
	}
 
	wk=(double *) calloc((size_t) (3*ndb), sizeof(double));
	wk1=(double *) calloc((size_t) (3*ndb), sizeof(double));
/*     Set up the matrix for equations */
	for(i=0; i<nx; ++i) {
		xb=x[i];
		ier=bsplin(xf,mx,k,xb,nderiv,wk,&wk[ndb],&wk[2*ndb],&left);
		if(ier>100) {free(wk1); free(wk); return ier;}
		for(j=0; j<ny; ++j) {
			yb=y[j];
			ier=bsplin(yf,my,k,yb,nderiv,wk1,&wk1[ndb],&wk1[2*ndb],&left);
			if(ier>100) {free(wk1); free(wk); return ier;}
			if(ef[i+j*ic]<= 0.0) {free(wk1); free(wk); return 610;}
			ij=i+j*nx;
			for(k1=0; k1<m1; ++k1) {
				for(k2=0; k2<m2; ++k2) {
					ij1=k1+k2*m1;
					a[ij1+ij*la]=wk[k1]*wk1[k2]/ef[i+j*ic];
					if(rlm>0.0) {
						a[ij1+(ij+nx*ny)*la]=rlm*wk[nb+k1]*wk1[k2];
						a[ij1+(ij+2*nx*ny)*la]=rlm*wk[k1]*wk1[nb+k2];
					}
				}
			}
		}
	}
	free(wk1); free(wk);

	ier=svd(m,n1,a,v,sigma,la,iv);
	if(ier>100)  return ier;

	wk=(double *) calloc((size_t) n1, sizeof(double)); 
/*     Setup the RHS and solve the equations */
	for(i=0; i<nx; ++i) {
		for(j=0; j<ny; ++j) {
			ij=i+j*nx;
			wk[ij]=f[i+j*ic]/ef[i+j*ic];
			if(rlm>0.0) {
				wk[ij+nx*ny]=0.0;
				wk[ij+2*nx*ny]=0.0;
			}
		}
	}

	ier=svdevl(m,n1,a,v,sigma,la,iv,wk,reps);

	for(i=0; i<m1; ++i) {
		for(j=0; j<m2; ++j) c[i+j*ic]=wk[i+m1*j];
	}
	free(wk);
 
/*     Calculate the \chi^2 */
	*chisq=0.0;
	nderiv=0;
	for(i=0; i<ny; ++i) {
		for(j=0; j<nx; ++j) {
			fy[j+i*ic]=bspev2(mx,my,xf,yf,k,nderiv,c,ic,x[j],y[i],&dfx,
					&dfy,&dfxx,&dfxy,&dfyy,&ier);
			r1=(f[j+i*ic]-fy[j+i*ic])/ef[j+i*ic];
			*chisq=(*chisq)+r1*r1;
		}
	}
	return 0;
}

File: projects/antia_numerical_methods@Beliavsky/f77/sub8/bspfitw2.f
!	To calculate linear least squares fit to B-spline basis functions in 2 dimension
!	version for general weights, but much slower than BSPFIT2
!
!	NX : (input) Number of data points along x-axis
!	NY : (input) Number of data points along y-axis
!	X,Y : (input) Real array of length NX,NY containing the coordinates
!		of points at which function values is available
!	F : (input) Real array of length IC*NY containing the function values
!		F(I,J) should be function value at X(I),Y(J)
!	EF : (input) Real array of length IC*NY containing the error estimates in F(I,J)
!	K : (input) Order of B-splines required, K=4 gives cubic B-splines
!	A : (output) Real array of length LA*(MX+K-2)*(MY+K-2) containing
!		the matrix U of SVD of the design matrix
!	LA : (input) First dimension of array A as declared
!		in the calling program (LA .GE. 3*NX*NY)
!	V : (output) Real array of length IV*(MX+K-2)*(MY+K-2) containing
!		the matrix V of SVD of the design matrix
!	IV : (input) First dimension of V in the calling program
!		IV .GE. (MX+K-2)*(MY+K-2)
!	SIGMA : (output) Real array of length (MX+K-2)*(MY+K-2)
!		containing the singular values of the design matrix
!	C : (output) Real array of length IC*(MY+K-2) containing the
!		fitted coefficients
!	IC : (input) First dimension of arrays C, F, EF, FY as declared
!		in the calling program (IC .GE. NX)
!	XF : (input) Real array of size MX, containing the knots
!		along x-axis used for defining B-spline basis functions.
!	YF : (input) Real array of size MY, containing the knots
!		along y-axis used for defining B-spline basis functions.
!	MX : (input) Number of knots for B-splines along x-axis,
!		the number of basis functions would be MX+K-2
!	MY : (input) Number of knots for B-splines along y-axis,
!		the number of basis functions would be MY+K-2
!	FY : (output) Real array of length IC*NY containing the values of fitted
!		function at each of the tabular points
!	WK : Real array of length 3*NX*NY+(MX+K)*(MY+K) used as scratch space
!	REPS : (input) Required accuracy for solution of equations using SVD
!		singular values less than REPS times maximum will be set to zero
!	RLM : (input) Parameter lambda for smoothing. If RLM.LE.0 no smoothing
!		is applied
!	IDE : (input) Order of derivative to be used for smoothing
!		This is used only when RLM>0. IDE=1 for first derivative
!		and IDE=2 for second derivative smoothing
!	CHISQ : (output) The value of Chi square at minimum
!	IER : (output) Error parameter, IER=0 implies successful execution
!		IER=608 implies that MX+K-2>NX, MY+K-2>NY or K<2
!		IER=609 implies that RLM>0 and IDE is not acceptable
!		IER=610 implies that EF(I,J).LE.0 for some I,J
!		No calculations are done in all these cases
!		Other values of IER may be set by SVD or BSPLIN
!
!	Required routines : BSPLIN, BSPEV2, SVD, SVDEVL
!
      SUBROUTINE BSPFITW2(NX,NY,X,Y,F,EF,K,A,LA,V,IV,SIGMA,
     1      C,IC,XF,YF,MX,MY,FY,WK,REPS,RLM,IDE,CHISQ,IER)
      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION X(NX),Y(NY),A(LA,*),XF(MX),YF(MY),C(IC,*),F(IC,NY),
     1      EF(IC,NY),V(IV,*),SIGMA(*),FY(IC,NY),WK(*)
 
      IF(NX.LT.K.OR.NY.LT.K.OR.K.LT.2) THEN
        IER=608
        RETURN
      ENDIF
 
      IF(RLM.GT.0.0.AND.(IDE.LT.1.OR.IDE.GT.2)) THEN
        IER=609
        RETURN
      ENDIF
!     N1 is the number of equations to be solved
      N1=NX*NY
      IF(RLM.GT.0.0) N1=3*N1
 
!     Set up the matrix equation and obtain SVD
!     M is the number of coefficients to be determined
      M1=MX+K-2
      M2=MY+K-2
      M=M1*M2
      NDB=MAX(MX,MY)+5
      NDERIV=0
      IF(RLM.GT.0.0) THEN
        NDERIV=IDE
        NB=NDB-1
        NB1=NDB*4-1
        IF(IDE.EQ.2) NB=2*NDB-1
        IF(IDE.EQ.2) NB1=5*NDB-1
      ENDIF
 
!     Set up the matrix for equations
      DO 2500 I=1,NX
        XB=X(I)
        CALL BSPLIN(XF,MX,K,XB,NDERIV,WK,WK(NDB),WK(NDB*2),LEFT,IER,
     1              WK(NDB*3))
        IF(IER.GT.100) RETURN
        DO 2500 J=1,NY
          YB=Y(J)
          CALL BSPLIN(YF,MY,K,YB,NDERIV,WK(NDB*3),WK(NDB*4),WK(NDB*5),
     1                LEFT,IER,WK(NDB*6))
          IF(IER.GT.100) RETURN
          IF(EF(I,J).LE.0) THEN
            IER=610
            RETURN
          ENDIF
          IJ=I+(J-1)*NX
          DO 2200 K1=1,M1
            DO 2200 K2=1,M2
              IJ1=K1+(K2-1)*M1
              A(IJ,IJ1)=WK(K1)*WK(NDB*3+K2-1)/EF(I,J)
              IF(RLM.GT.0) A(IJ+NX*NY,IJ1)=RLM*WK(NB+K1)*WK(NDB*3+K2-1)
              IF(RLM.GT.0.0) A(IJ+2*NX*NY,IJ1)=RLM*WK(K1)*WK(NB1+K2)
2200      CONTINUE
2500  CONTINUE
      CALL SVD(M,N1,A,V,SIGMA,LA,IV,WK,IER)
      IF(IER.GT.100) RETURN
 
 
 
!     Setup the RHS and solve the equations
      DO 3000 I=1,NX
        DO 3000 J=1,NY
          IJ=I+(J-1)*NX
          WK(IJ)=F(I,J)/EF(I,J)
          IF(RLM.GT.0.0) WK(IJ+NX*NY)=0.0
          IF(RLM.GT.0.0) WK(IJ+2*NX*NY)=0.0
3000  CONTINUE
 
      CALL SVDEVL(M,N1,A,V,SIGMA,LA,IV,WK,WK(N1+2),REPS)
 
      DO 3400 I=1,M1
        DO 3400 J=1,M2
          C(I,J)=WK(I+M1*(J-1))
3400  CONTINUE
 
 
!     Calculate the \chi^2
      CHISQ=0.0
      NDERIV=0
      DO 4000 I=1,NY
        DO 4000 J=1,NX
          FY(J,I)=BSPEV2(MX,MY,XF,YF,K,NDERIV,C,IC,X(J),Y(I),DFX,DFY,
     1                   DFXX,DFXY,DFYY,WK,IER1)
          CHISQ=CHISQ+((F(J,I)-FY(J,I))/EF(J,I))**2
4000  CONTINUE
 
      END
